woocommerce/woocommerce-ios
# Attempt mitigate tablet crash when `highlightSelectedRowIfNeeded`

Closes: #12374

## Description

This PR attempts to mitigate a crash we've seen on tablets and described on peaMlT-vd-p2. When tapping in Orders, there seems to be a crash in some instances when the system attempts to highlight the appropriate row for the selected order.

> An initial look seems to be from converting Storage.Order (NSManagedObject)’s items: NSOrderedSet of type Storage.OrderItem to a Swift array. Something went wrong in the conversion from NSOrderedSet‘s Objective-C element to the read-only Swift version.

Since I couldn't replicate it, my initial approach to this is to be more strict around the type conversion, and to move the `highlightSelectedRowIfNeeded` method to be called only when needed. Any input or feedback is welcome.

## Changes:
1. Moved `highlightSelectedRowIfNeeded()` from `willDisplay` to `viewDidAppear`

While this delegate allows us to customize the cell before it is displayed and highlight if its needed, we're calling this method multiple times when is not necessary (eg: when scrolling). We only need one check when the view is loaded, and further checks will happen when the different rows are selected. 

Additionally, internally `highlightSelectedRowIfNeeded` is reaching to storage to calculate the index path each time this delegate was called, which could be causing concurrency issues if we're accessing to the NSOrderedSet from multiple threads or there's some sort of index path miss-match since cells should've been reorganized.

```
EXC_BAD_ACCESS: Exception 1, Code 2, Subcode 6095503072 >
Stack overflow in (null)
  File "NSOrderedSet+Array.swift", line 5, in NSOrderedSet.toArray<T>
  File "Order+ReadOnlyConvertible.swift", line 166, in Order.orderItemsArray.getter
  File "Order+ReadOnlyConvertible.swift", line 74, in Order.toReadOnly
  File "<compiler-generated>", in Order
  File "FetchResultSnapshotsProvider.swift", line 175, in FetchResultSnapshotsProvider.object
...
(103 additional frame(s) were not displayed)
```

2. Added a `toTypeCheckedArray` method

Despite using optional casting during type conversion, we've seen some logs pointing to force casting and bridging issues between types. Eg:
```
Array._unconditionallyBridgeFromObjectiveC, 
_arrayForceCast<T>
```

This is a copy of `.toArray()` with additional safeguards by being stricter catching type-cast errors, since we're checking each element rather than the entire array level, and provides feedback about why the cast failed via the `assert`. If any of the elements cannot be casted we'll just return an empty array in production.

3. Updates `orderItemsArray` nil-coalescing fallback type to be explicit

When the read-only convertible implementation fails to convert  to a Swift array, we fallback to an empty array `[]` which will be type-inferred automatically. There shouldn't be any difference between being explicit or not, but since we've seen from the logs that there might be errors when casting/converting between types, we can be explicit here and just tell the compiler and the next person that we're expecting an empty `[Storage.OrderItem]`

## Testing instructions
I wasn't able to replicate the issue on peaMlT-vd-p2 . My attempts to load a site with a large number of orders that hasn’t been synchronised for a while seems to just make the selected row to jump "out of the view" when the rest of cells are updated, but there's no crash.

Run the app on tablet, go to Orders, and test that the selected row works as expected on selection, deselection, moving between views and playing with the split view configuration.
